File: SampleApps/ObjC/SampleBlueDotExperience/SampleBlueDotExperience/UI/BlueDot/ViewController.m
//
//  ViewController.m
//  SampleBlueDotWithIndoorLocation
//
//  Created by Akhil Rana on 24/11/23.
//

#import "ViewController.h"
#import "Constants.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIView *mapContainer;
@property (strong, nonatomic) UIView *blueDot;

@property (strong, nonatomic) MistService *mistService;

@property CGFloat scaleX;
@property CGFloat scaleY;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setup];
    [_mistService start];
 }

- (void)setup {
    _blueDot = [self createBlueDot];
    self.mistService = [[MistService alloc] initWithToken: SDK_TOKEN];
    self.mistService.delegate = self;
 }

- (void)didUpdateLocation:(CGPoint)location {
    CGFloat scaledX = location.x * _scaleX;
    CGFloat scaledY = location.y * _scaleY;
    CGPoint relativeLocation = CGPointMake(scaledX, scaledY);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.blueDot.center = relativeLocation;
    });
 }

- (void) didUpdateMap:(NSURL *)mapURL {
    [_mistService downloadImageWithURL:mapURL completion:^(UIImage * _Nullable image, NSError * _Nullable error) {
        if (error) {
            NSLog(@"Error downloading image: %@", error.localizedDescription);
        } else {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self setupMapViewWithImage:image];
            });
        }
    }];
 }

- (UIView *)createBlueDot {
    if (!_blueDot) {
        _blueDot = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 10, 10)];
        _blueDot.translatesAutoresizingMaskIntoConstraints = NO;
        _blueDot.layer.cornerRadius = 5;
        _blueDot.backgroundColor = [UIColor colorWithRed:0.072 green:0.593 blue:0.997 alpha:1.0];
    }
    return _blueDot;
}

- (void)setupMapViewWithImage:(UIImage *)image {
    UIView *floorMapView = [[UIView alloc] initWithFrame:self.view.bounds];
    floorMapView.translatesAutoresizingMaskIntoConstraints = NO;
    floorMapView.backgroundColor = UIColor.whiteColor;
    [self.mapContainer addSubview:floorMapView];

    UIImageView *floorImageView = [[UIImageView alloc] initWithImage:image];
    floorImageView.contentMode = UIViewContentModeScaleAspectFit;
    floorImageView.translatesAutoresizingMaskIntoConstraints = NO;
    [floorMapView addSubview:floorImageView];

    [floorMapView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor].active = YES;
    [floorMapView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor].active = YES;
    [floorMapView layoutIfNeeded];

    CGFloat floorViewRatio = self.mapContainer.bounds.size.width / self.mapContainer.bounds.size.height;
    CGFloat imageRatio = image.size.width / image.size.height;

    if (imageRatio >= floorViewRatio) {
        [floorMapView.widthAnchor constraintEqualToAnchor:self.mapContainer.widthAnchor constant:0].active = YES;
        [floorMapView.heightAnchor constraintEqualToAnchor:self.mapContainer.widthAnchor multiplier:image.size.height / image.size.width constant:0].active = YES;
    } else {
        [floorMapView.widthAnchor constraintEqualToConstant:self.mapContainer.bounds.size.height].active = YES;
        [floorMapView.heightAnchor constraintEqualToAnchor:self.mapContainer.widthAnchor multiplier:image.size.width / image.size.height constant:0].active = YES;
    }

    [floorMapView layoutIfNeeded];

    [floorMapView addSubview:self.blueDot];
    [floorMapView bringSubviewToFront:self.blueDot];

    [NSLayoutConstraint activateConstraints:@[
        [floorImageView.topAnchor constraintEqualToAnchor:floorMapView.topAnchor],
        [floorImageView.leadingAnchor constraintEqualToAnchor:floorMapView.leadingAnchor],
        [floorImageView.trailingAnchor constraintEqualToAnchor:floorMapView.trailingAnchor],
        [floorImageView.bottomAnchor constraintEqualToAnchor:floorMapView.bottomAnchor],

        [self.blueDot.widthAnchor constraintEqualToConstant:10.0],
        [self.blueDot.heightAnchor constraintEqualToConstant:10.0]
    ]];

    [self.mapContainer layoutIfNeeded];

    // Calculate the Scale Factor
    _scaleX = floorMapView.bounds.size.width / image.size.width;
    _scaleY = floorMapView.bounds.size.height / image.size.height;
}

@end

File: Sample Apps /SampleBlueDotWithIndoorLocation/SampleBlueDotWithIndoorLocation/UI/BlueDot/ViewController.swift
//
//  ViewController.swift
//  SampleBlueDotWithIndoorLocation
//
//  Created by Rajesh Vishwakarma on 27/01/23.
//

import UIKit

class ViewController: UIViewController {
    
    @IBOutlet private var wakeUpSwitch: UISwitch!
    @IBOutlet private var mistSdkSwitch: UISwitch!
    @IBOutlet private var mapContainer: UIView!
    @IBOutlet private var statusLabel: UILabel!
    
    private let blueDot: UIView = {
        let dotView = UIView(frame: .init(x: 0, y: 0, width: 10, height: 10))
        dotView.translatesAutoresizingMaskIntoConstraints = false
        dotView.layer.cornerRadius = 5
        dotView.backgroundColor = UIColor(red: 0.072, green: 0.593, blue: 0.997, alpha: 1.0)
        return dotView
    }()
    
    var viewModel: ViewModelDelegate?
    
    private var scale: Scale?
    private var isMapLoaded = false
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        
        // Assemble Component and Inject Dependencies
        configurator()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        statusLabel.isHidden = true
        
        wakeUpSwitch.isOn = viewModel?.isMistServiceRuning ?? false
        mistSdkSwitch.isOn = viewModel?.isMistServiceRuning ?? false
    }
    
    @IBAction func enableMistSDKService(_ sender: UISwitch) {
        
        if sender.isOn {
            HUDProgress.shared.start()
            viewModel?.startMistService()
        } else {
            viewModel?.stopMistService()
        }
    }
    
    @IBAction func enableAppWakeUpService(_ sender: UISwitch) {
        
        if sender.isOn {
            viewModel?.startWakeUp()
        } else {
            viewModel?.stopWakeUp()
        }
    }
    
    func updateStatus(_ loc: CGPoint) {
        let ppm = viewModel?.mapPPMValue ?? 1
        let clientLocation = CGPoint(x: loc.x / ppm, y: loc.y / ppm)
        statusLabel.text = String(format: "BlueDot Location on Map X= %.1f, Y= %.1f", clientLocation.x, clientLocation.y)
    }
}

extension ViewController {
    
    private func configurator() {
        #if !targetEnvironment(simulator)
        let mistService = RealMistService(token: Mist.SDK.token)
        let wakeUpService = RealWakeupService()
        let viewModel = ViewModel(mistService: mistService, wakeUpService: wakeUpService)
        mistService.delegate = viewModel
        wakeUpService.delegate = viewModel
        viewModel.delegate = self
        self.viewModel = viewModel
        #endif
    }
    
    private func setupMapView(with image: UIImage) {
        
        let floorMapView = UIView(frame: self.view.bounds)
        floorMapView.translatesAutoresizingMaskIntoConstraints = false
        floorMapView.backgroundColor = .white
        mapContainer.addSubview(floorMapView)
        
        let floorImageView = UIImageView(image: image)
        floorImageView.contentMode = .scaleAspectFit
        floorImageView.translatesAutoresizingMaskIntoConstraints = false
        floorMapView.addSubview(floorImageView)
        
        floorMapView.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        floorMapView.centerYAnchor.constraint(equalTo: view.centerYAnchor).isActive = true
        floorMapView.layoutIfNeeded()
        
        let floorViewRatio = mapContainer.bounds.size.width / mapContainer.bounds.size.height
        let imageRatio = image.size.width/image.size.height
        
        if imageRatio >= floorViewRatio {
            floorMapView.widthAnchor.constraint(equalTo: mapContainer.widthAnchor, constant: 0).isActive = true
            floorMapView.heightAnchor.constraint(equalTo: mapContainer.widthAnchor, multiplier: image.size.height/image.size.width, constant: 0).isActive = true
        } else {
            floorMapView.widthAnchor.constraint(equalToConstant: mapContainer.bounds.height).isActive = true
            floorMapView.heightAnchor.constraint(equalTo: mapContainer.widthAnchor, multiplier: image.size.width/image.size.height, constant: 0).isActive = true
        }
        
        floorMapView.layoutIfNeeded()
        
        floorMapView.addSubview(blueDot)
        floorMapView.bringSubviewToFront(blueDot)
        
        NSLayoutConstraint.activate([
            floorImageView.topAnchor.constraint(equalTo: floorMapView.topAnchor),
            floorImageView.leadingAnchor.constraint(equalTo: floorMapView.leadingAnchor),
            floorImageView.trailingAnchor.constraint(equalTo: floorMapView.trailingAnchor),
            floorImageView.bottomAnchor.constraint(equalTo: floorMapView.bottomAnchor),
            
            blueDot.widthAnchor.constraint(equalToConstant: 10.0),
            blueDot.heightAnchor.constraint(equalToConstant: 10.0)
        ])
        
        mapContainer.layoutIfNeeded()
        
        // Calculate the Scale Factor
        let scaleX = floorMapView.bounds.width / image.size.width
        let scaleY = floorMapView.bounds.height / image.size.height
        
        scale = Scale(x: scaleX, y: scaleY)
    }
}


// MARK: - ViewDelegate

extension ViewController: ViewDelegate {
    
    func didMistServiceStarted() {
        HUDProgress.shared.start()
        if let mistSdkSwitch = mistSdkSwitch {
            mistSdkSwitch.isOn = true
        }
    }
    
    func didLoadMistMap(with image: UIImage) {
        HUDProgress.shared.stop()
        setupMapView(with: image)
        isMapLoaded = true
        statusLabel.isHidden = false
    }
    
    func didUpdateMistLocation(with point: CGPoint) {
        guard isMapLoaded, let scale = scale else { return }
        let relativeLocation = point.scaleUpPoint(scale: scale)
        self.blueDot.center = relativeLocation
        updateStatus(point)
    }
}

